File: code/task2predict.py
from pyspark import SparkContext
from math import sqrt
import json
import sys
import time


class T2p:
    def __init__(self) -> None:
        self.testfile = sys.argv[1]
        self.modelfile = sys.argv[2]
        self.outputfile = sys.argv[3]

    @staticmethod
    def cosine_sim(uservec, bizvec):
        if not uservec or not bizvec:
            return 0
        return len(uservec & bizvec) / (sqrt(len(uservec)) * sqrt(len(bizvec)))

    def run(self):
        sc = SparkContext.getOrCreate()
        sc.setLogLevel("OFF")
        
        modelRDD = sc.textFile(self.modelfile).map(lambda row: json.loads(row))

        uzp = {x["user_id"] : set(x["fvec"]) for x in modelRDD.filter(lambda row: row["type"] == "user").collect()}
        bzp = {x["biz_id"] : set(x["fvec"]) for x in modelRDD.filter(lambda row: row["type"] == "biz").collect()}
        
        ans = sc.textFile(self.testfile).map(lambda row: json.loads(row)).map(lambda row: (row["user_id"], row["business_id"])).map(lambda x: (x[0], x[1], T2p.cosine_sim(uzp.get(x[0]), bzp.get(x[1])))).filter(lambda x: x[2] >= 0.01).collect()
        
        with open(self.outputfile, "w+") as f:
            for row in ans:
                f.write(json.dumps({"user_id": row[0], "business_id": row[1], "sim": row[2]}) + "\n")


if __name__ == "__main__":
    t2p = T2p()

    st = time.time()
    t2p.run()
    et = time.time()
    print("Total time taken: {}".format(et - st))

File: scala/task2predict.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.SparkContext
import org.apache.spark.rdd.RDD
import org.apache.spark.storage.StorageLevel
import org.json4s._
import org.json4s.jackson.JsonMethods._

import scala.math.log10
import scala.util.control.Breaks._
import scala.util.Random._
import collection.mutable.{ArrayBuffer, HashMap, HashSet}
import java.io.{File, PrintWriter}
import scala.io.Source
import scala.math.Ordering.Implicits._
import scala.math.sqrt
import scala.util.Random


object task2predict {
  implicit val formats = DefaultFormats

  def cosineSim(userVec: Set[String], bizVec: Set[String]): Double = {
    if (userVec.size == 0)
      return 0
    if (bizVec.size == 0)
      return 0

    val num = userVec.intersect(bizVec).size
    val den = sqrt(userVec.size) * sqrt(bizVec.size)

    return num.toDouble / den.toDouble
  }

  def writeData(opf: String, data: Array[(String, String, Double)]) = {
    val out = new PrintWriter(new File(opf))

    for (item <- data) {
      val output: Map[String, Any] = Map("user_id"->item._1, "business_id"->item._2, "sim"->item._3)
      val formatted_output = org.json4s.jackson.Serialization.write(output)
      out.write(formatted_output)
      out.write("\n")
    }

    out.close()
  }

  def main(args: Array[String]) = {
    val t1 = System.nanoTime

    val sc = SparkContext.getOrCreate()
    sc.setLogLevel("OFF")
    val testfile = args(0)
    val modelfile = args(1)
    val outputfile = args(2)

    val modelRDD = sc.textFile(modelfile).map(row => parse(row))

    val uzp = modelRDD.filter(row => (row\"type").extract[String] == "user")
      .map(row => ((row\"user_id").extract[String], (row\"fvec").extract[Array[String]]))
      .mapValues(value => value.toSet)
      .collectAsMap()

    val bzp = modelRDD.filter(row => (row\"type").extract[String] == "biz")
      .map(row => ((row\"biz_id").extract[String], (row\"fvec").extract[Array[String]]))
      .mapValues(value => value.toSet)
      .collectAsMap()

    val ans = sc.textFile(testfile)
      .map(row => parse(row))
      .map(row => ((row\"user_id").extract[String], (row\"business_id").extract[String]))
      .map(row => (row._1, row._2, cosineSim(uzp(row._1), bzp(row._2))))
      .filter(row => row._3 >= 0.01)
      .collect()

    writeData(outputfile, ans)

    println("Duration: " + (System.nanoTime - t1) / 1e9d)
  }
}
